Add type annotations

miedzinski opened the pull request:
Another take on adding type annotations to `bottom`.

There are few `# type: ignore` comments, but that's okay:

- currently mypy doesn't consider `functools.partial` a subtype of `Callable`; that will be fixed when structural subtyping is implemented
- awaiting `AbstractEventLoop.create_connection` call - I have reported it on python/typeshed#953.
- signature override in `protocol.Protocol.connection_made`, because we need both `ReadTransport` and `WriteTransport`. If you want, I can rewrite that method to look like this

```python
    def connection_made(self, transport: asyncio.BaseTransport) -> None:
        assert isinstance(transport, asyncio.Transport)
        self.transport = transport
```

and we'll have the benefit of being sure that right transport is being used, and mypy will be happy.

Besides that everything is cleanly typechecking.

Currently there is `stubs` directory with annotations for `simplex`. If you're okay with this I'd remove this and add typing to `simplex` repository as well (with py2-compatible syntax).

miedzinski commented:
I have changed one type annotation to be more specific and removed one `# type: ignore` comment with changes suggested in PR description.

If you decide to merge this, please decide what to do with `simplex` first. If you want to add type annotation there, tell me whether they should be py2 compatible (function annotations or type comments).

numberoverzero commented:
Thanks for the PR!  I'm +1 for type annotations, but there are a few small artifacts that I'm not a fan of.  I'll try to add concrete feedback over the weekend, if things don't come up.

Every time it comes up I dislike the `simplex` dependency more and more.  It's personally useful, but it isn't core to what bottom offers and really is more of an example of using the plugin system.  I'd prefer to pull it out entirely, bump major version, drop `simplex` from requirements, and put it in the Extensions/Plugins docs.  Looks like this work would also benefit from that effort.

miedzinski commented:
Sure. I dislike `simplex` too; `bottom` is quite low-level IRC library (and this is one of the main reasons I love it), but for some reason there is something like `Router`.

miedzinski commented:
@numberoverzero, done. I've changed `isinstance(..., Transport)` assertion to `WriteTransport` - we don't need `ReadTransport`. And since mypy 0.501 was released few days ago, I've changed it's version in tox config to PyPI one. There are still few `# type: ignore`s; the `functools.partial` one needs python/typing#11 (which won't land too soon) and `AbstractEventLoop.create_connection`, which already has PR ready in typeshed repo (and will hopefully get included in next mypy release).

numberoverzero commented:
Thanks!  Taking a look.

numberoverzero commented on bottom/client.py:
```diff
@@ -1,14 +1,20 @@
+    _events = None  # type: DefaultDict[str, asyncio.Event]
+
+    def __init__(self, host: str, port: int, *,
+                 encoding: str = "UTF-8", ssl: bool = True,
```
Does PEP 484 recommend spaces/no spaces around the equals sign for default values?  Without types I've used `some_param=None` so I'd expect this to be `some_param: Optional[Thing]=None`.

miedzinski replied:
It doesn't strictly recommend it, but examples use this style. Mypy codebase uses this as well. I can change it, but I think the community will stick to version with spaces (given annotation is present).

numberoverzero commented on bottom/client.py:
```diff
@@ -1,14 +1,20 @@
 import asyncio
 import collections
 import functools
+from typing import Any, Callable, DefaultDict, List, Optional  # noqa
```
`DefaultDict` was added after 3.5.0, so bottom won't be strictly 3.5+.  However, upgrading patch versions and even 3.4 -> 3.5, 3.5 -> 3.6 have been relatively painless (though adoption lags, see RTD build environments).

How important is this over just using `Dict`?  Will the changes for `functools.partial` and `AbstractEventLoop` require a new python release, or will they use existing types?

miedzinski replied:
Using `Dict` here is perfectly fine - I will add a commit. `functools.partial` (protocols in general) will probably need new `typing` module, so it may need a new release. However, in this case we aren't defining any new protocols, but using builtin class, so future mypy version will simply use new `typing.pyi` from typeshed and it will be sufficient. Same goes for `AbstractEventLoop` methods: once this is fixed in typeshed repo, it will be merged into mypy and then released on PyPI; no new Python distribution needed.

numberoverzero commented on bottom/protocol.py:
```diff
@@ -1,29 +1,38 @@
 from bottom.unpack import unpack_command
+from typing import Optional, TYPE_CHECKING
+
+if TYPE_CHECKING:
```
What other options are there instead of the circular import just to get a reference to `Client`?  I don't like circular imports[0] (module-only or otherwise) and I'll use `Any` before I add the headache of tracing imports when I refactor code.  A stubs module?  Not a huge fan but I could at least stomach that.

[0] I'm still disappointed that PEP 484 so casually recommends circular imports for something core to the type annotations problem space.

miedzinski replied:
`typing.TYPE_CHECKING` is always `False` at runtime, so there isn't actually any circular import. This flag is seen as `True` by mypy (and other typecheckers) and as we can see it handles this case pretty well. There has to be some kind of import for mypy to follow; otherwise it wouldn't know what `Client` actually is (well, in this codebase there is only one `Client`, but I hope you get the point). It's fine.

numberoverzero commented on bottom/protocol.py:
```diff
@@ -1,29 +1,38 @@
+    def __init__(self, client: 'Optional[Client]' = None) -> None:
+        if client is not None:
+            self.client = client
+        self.closed = False  # type: bool
```
Can `closed` and `buffer` move up to the class?  `bool` and `bytes` won't copy across instances.

miedzinski replied:
Are you sure? If they weren't defined in `__init__`, all instances would share the same objects, which could cause bugs.

numberoverzero commented on bottom/protocol.py:
```diff
@@ -1,29 +1,38 @@
+    transport = None  # type: asyncio.WriteTransport
+
+    def __init__(self, client: 'Optional[Client]' = None) -> None:
+        if client is not None:
```
Why isn't this just `self.client = client`?  That's the behavior we'd get anyway.

miedzinski replied:
We'd explicitly initialize `self.client` with `Optional[Client]` value, but the variable is `Client`. We can

a) make `Protocol.client` `Optional` and do checks in all methods... I don't like it
b) let `Protocol.__init__` accept `Client` instead of `Optional[Client]` - this would require rewriting some text fixtures, right?
c) leave it as is :)

numberoverzero commented:
Thanks very much!  If you haven't yet, feel free to add yourself to the Contributors section in the README.
